document_search.go

```diff
@@ -0,0 +1,262 @@
+
+	// String implements SearchValue that holds a string value
+	String struct {
+		Value string
+	}
+
+	// Float64 implements SearchValue that holds a float64 value
+	Float64 struct {
+		Value float64
+	}
+
+	// Int implements SearchValue that holds an int value
+	Int struct {
+		Value int
+	}
+
+	// Time implements SearchValue that holds an time.Time value
+	Time struct {
+		Value time.Time
+	}
+
+	// RangeValue implements SearchValue that holds range values
+	RangeValue struct {
+		Gt  float64 `json:"gt,omitempty"`
+		Gte float64 `json:"gte,omitempty"`
+		Lt  float64 `json:"lt,omitempty"`
+		Lte float64 `json:"lte,omitempty"`
+	}
+)
+
+func (i String) searchValue()     {}
+func (i Float64) searchValue()    {}
+func (i Int) searchValue()        {}
+func (i Time) searchValue()       {}
+func (i RangeValue) searchValue() {}
+
+// MarshalJSON ...
+func (i String) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(i.Value)
+}
+
+// MarshalJSON ...
+func (i Float64) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(i.Value)
+}
+
+// MarshalJSON ...
+func (i Int) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(i.Value)
+}
+
+// MarshalJSON ...
+func (i Time) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(i.Value)
+}
+
+type (
+	// SearchType ...
+	SearchType interface {
+		searchType()
+	}
+
+	// SearchTypes ...
+	SearchTypes []SearchType
+
+	// Eq ...
+	Eq struct {
+		Value         SearchValue
+		CaseSensitive bool
+	}
+
+	// In ...
+	In struct {
+		Value         []SearchValue
+		CaseSensitive bool
+	}
+
+	// Not ...
+	Not struct {
+		Value         SearchValue
+		CaseSensitive bool
+	}
+
+	// NotIn ...
+	NotIn struct {
+		Value         []SearchValue
+		CaseSensitive bool
+	}
+
+	// Wildcard ...
+	Wildcard struct {
+		Value         SearchValue
+		CaseSensitive bool
+	}
+
+	// Range ...
+	Range struct {
+		Value RangeValue
+	}
+)
+
+func (i In) searchType()       {}
+func (i Eq) searchType()       {}
+func (i Not) searchType()      {}
+func (i NotIn) searchType()    {}
+func (i Wildcard) searchType() {}
+func (i Range) searchType()    {}
+
+// MarshalJSON ...
+func (i In) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(struct {
+		Type          string        `json:"type,omitempty"`
+		Value         []SearchValue `json:"value,omitempty"`
+		CaseSensitive bool          `json:"case_sensitive,omitempty"`
+	}{
+		Type:          "in",
+		Value:         i.Value,
+		CaseSensitive: i.CaseSensitive,
+	})
+}
+
+// MarshalJSON ...
+func (i Eq) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(struct {
+		Type          string      `json:"type,omitempty"`
+		Value         SearchValue `json:"value,omitempty"`
+		CaseSensitive bool        `json:"case_sensitive,omitempty"`
+	}{
+		Type:          "eq",
+		Value:         i.Value,
+		CaseSensitive: i.CaseSensitive,
+	})
+}
+
+// MarshalJSON ...
+func (i Not) MarshalJSON() (data []byte, err error) {
+	return json.Marshal(struct {
```

Comment:
why use an anonymous struct in all of these? Seems like lots of duplication

Reply:
some appear to be different

Reply:
yeah i decided making anonymous structs for that reason. Sometimes the json dont have case_senstive property or the value is a list or just a single value.

Reply:
used named structs here https://github.com/FirstVisit/go-truevault/pull/1/commits/5a4d6cdf3382b5172b62225c60af36d5dd93775c